### ISSUE TITLE
Fix trust region exit criterion

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#1067]
In  ```trust_region.m``` we have a problem with trying to encode three pieces of information in a binary variable ```check```

The problem is error code -3 (Trust region became uselessly small). c01352ea3bfd6dd29074a2e0867afc5a1763f79d tried to prevent the solver from accepting error code -3 as a solved steady state. But this breaks the loop in ```dynare_solve.m```:

```
    for i=length(r)-1:-1:1
        if options.debug
            disp(['DYNARE_SOLVE (solve_algo=2|4): solving block ' num2str(i) ', of size ' num2str(r(i+1)-r(i)) ]);
        end
        [x,info]=solver(func,x,j1(r(i):r(i+1)-1),j2(r(i):r(i+1)-1),jacobian_flag, ...
                        options.gstep, ...
                        tolf,options.solve_tolx, ...
                        options.steady.maxit,options.debug,varargin{:});
        if info
            return
        end
    end
```
which relies on being able to continue with the loop when the code internally was -3 (as ~(-3)=0), i.e. we discard the information that -3 is a problem. But further below we have the final check
```
    if max(abs(fvec)) > tolf
        [x,info]=solver(func,x,1:nn,1:nn,jacobian_flag, ...
                        options.gstep, tolf,options.solve_tolx, ...
                        options.steady.maxit,options.debug,varargin{:});
    end
```
Here, we would need to know whether a 0 in info comes from info = ~(1) (everything fine) vs. info=~(-3) (still a problem after the last iteration). 

My proposed solution would be to augment both ```trust_region.m``` and ```solve1.m``` with a third output argument ```exitflag``` that encodes the full integer error code. That way we would only need to check for this argument in the final check of ```dynare_solve.m```

**Comment by MichelJuillard** [comment]
I think that your first idea was the right one. If I understand correctly info can take three value at the end of the main loop
info = 0 : maxiter has been reached
info = -3: trust region too small
info = 1: success

The two first values indicate a problem and the procedure should stop and report failure

However, I think there was a problem in your first patch because when info == 1, check was returned undetermined instead of equal to 0

**Comment by JohannesPfeifer** [comment]
@MichelJuillard That was a different problem with my patch that does not solve the issue. The problem is that -3 is a non-terminal exit code within the loop, but it must be terminal outside of it. I pushed a proposed fix in #1068

**Comment by MichelJuillard** [comment]
I don't understand why "-3 is a non-terminal exit code within the loop"

**Comment by JohannesPfeifer** [comment]
Because in ```trust_region.m``` we have the line 
```
check = ~info;
```
So both 1 and -3 evaluate to check being 0, which is non-terminal for 
```
        if info
            return
        end
```

**Comment by MichelJuillard** [comment]
I think it is a bug. If the trust region becomes too small, it doesn't mean that we reached the solution

**Comment by JohannesPfeifer** [comment]
@MichelJuillard I think now that you are right. But the problem is that we do not check for a solution before breaking with code -3. I proposed an amended fix in #1068

**Comment by MichelJuillard** [comment]
@JohannesPfeifer, yes, it is better to check as you do now.
